src/github.com/getlantern/errlog/errlog.go

```diff
@@ -0,0 +1,540 @@
+	if e.Extra != nil && len(e.Extra) > 0 {
+		_, _ = buf.WriteString(",")
+		for k, v := range e.Extra {
+			_, _ = buf.WriteString(fmt.Sprintf(`"%s":"%s"`, k, v))
+		}
+	}
+	_, _ = buf.WriteString("}")
+	return buf.Bytes(), nil
+}*/
+
+type ErrorLogger struct {
+	goPackage string
+	logger    golog.Logger
+	*systemInfo
+}
+
+type withFunc func(e *Error)
+
+func WithOp(op string) withFunc {
+	return func(e *Error) {
+		e.Op = op
+	}
+}
+
+func WithProxy(info *ProxyingInfo) withFunc {
+	return func(e *Error) {
+		e.ProxyingInfo = info
+	}
+}
+
+func WithLocale() withFunc {
+	return func(e *Error) {
+		lang, _ := jibber_jabber.DetectLanguage()
+		country, _ := jibber_jabber.DetectTerritory()
+		e.UserLocale = &UserLocale{
+			time.Now().Format("MST"),
+			lang,
+			country,
+		}
+	}
+}
+
+func WithUserAgent(ua string) withFunc {
+	return func(e *Error) {
+		e.UserAgentInfo = &UserAgentInfo{ua}
+	}
+}
+
+func WithField(k, v string) withFunc {
+	return func(e *Error) {
+		e.Extra[k] = v
+	}
+}
+
+func (c *ErrorLogger) Log(err error, with ...withFunc) {
+	errOp, goType, desc, extra := parseError(err)
+	actual := &Error{
+		GoPackage:  c.goPackage,
+		GoType:     goType,
+		Desc:       desc,
+		Op:         errOp,
+		Extra:      extra,
+		systemInfo: c.systemInfo,
+	}
+	for _, f := range with {
+		f(actual)
+	}
+	currentReporter.Report(actual)
+	c.logger.Error(actual.String())
+}
+
+func ErrorLoggerFor(goPackage string) *ErrorLogger {
+	version, _ := osversion.GetHumanReadable()
+	return &ErrorLogger{
+		goPackage: goPackage,
+		logger:    golog.LoggerFor(goPackage),
+		systemInfo: &systemInfo{
+			OSType:    runtime.GOOS,
+			OSArch:    runtime.GOARCH,
+			OSVersion: version,
+		},
+	}
+}
+
+type Reporter interface {
+	Report(*Error)
+}
+
+var currentReporter Reporter = &StdReporter{}
+
+func ReportTo(r Reporter) {
+	currentReporter = r
+}
+
+func toJSON(e *Error) []byte {
+	b, err := json.Marshal(e)
+	if err != nil {
+		panic(fmt.Sprintf("failed to convert error to json: %+v", err))
+	}
+	return b
+}
+
+type StdReporter struct {
+}
+
+func (l StdReporter) Report(e *Error) {
+	fmt.Printf("%+v", string(toJSON(e)))
+}
+
+func parseError(err error) (op string, goType string, desc string, extra map[string]string) {
+	extra = make(map[string]string)
+
+	// interfaces
+	if _, ok := err.(net.Error); ok {
+		if opError, ok := err.(*net.OpError); ok {
+			op = opError.Op
+			if opError.Source != nil {
+				extra["localAddr"] = opError.Source.String()
+			}
+			if opError.Addr != nil {
+				extra["remoteAddr"] = opError.Addr.String()
+			}
+			extra["network"] = opError.Net
+			err = opError.Err
+		}
+		switch actual := err.(type) {
+		case *net.AddrError:
+			goType = "net.AddrError"
+			desc = actual.Err
+			extra["addr"] = actual.Addr
+		case *net.DNSError:
+			goType = "net.DNSError"
+			desc = actual.Err
+			extra["domain"] = actual.Name
+			if actual.Server != "" {
+				extra["dnsServer"] = actual.Server
```

Comment:
Rather than using camel-case for extras and JSON field names, how about using underscores like `dns_server`?  This data will be queried in InfluxDB and other databases, and the underscore convention is more common there.